### PR TITLE
add relative datediff macro (without tests)

### PR DESCRIPTION
**Pull request by jpmmcneill**
Hey @clausherther 

As discussed on slack - this is a macro I built that I found very helpful for things like customer lifetime analysis.

Let me know what you think!!!

### CODE REVIEW

**Comment by jpmmcneill** [comment]
I'd prefer for this to _not_ be merged until there is a macro for cross db date parts. I might submit a PR for that over the weekend :)

File: macros/utils/relative_datediff.sql [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    For example, consider dbt_utils.datediff('month', "'2021-09-01'", "'2021-10-30'")
+    This will return 1, despite there being almost 60 days in between the two.
+    This makes analysis quite difficul when you want to work in units like months or quarters,
+    but this should be derived from the number of days.
+
+    This macro works by looking at the date of date1, subtracting that (minus one) number of days to date 2
+    (which normalises date2 relative to date 1, ie. as if date 1 was the first of a month) and then computes
+    the normal date diff.
+
+    TODO: use a date_part from dbt utils (which doesn't exist right now :P) instead of hardcoding the function.
```
Comment (clausherther) [inline macros/utils/relative_datediff.sql]: Why not use `dbt_date.date_part()`?
https://github.com/calogica/dbt-date/blob/main/macros/calendar_date/date_part.sql

**Comment by clausherther** [comment]
> Hey @clausherther
> 
> As discussed on slack - this is a macro I built that I found very helpful for things like customer lifetime analysis.
> 
> Let me know what you think!!!

Looks interesting, thanks! I need to think about this works some more after coffee!

File: macros/utils/relative_datediff.sql (hunk 1 above)
Comment (jpmmcneill) [reply macros/utils/relative_datediff.sql]: Because I didn't know it existed 😂

Will do, thanks for highlighting :)

**Comment by jpmmcneill** [comment]
Marking as ready for review as I've put in the dbt_date.date_part, thanks for letting me know if existed 🤦

**Comment by clausherther** [comment]
Hi @jpmmcneill, thanks for this! I think we will also need an integration test for this. Let me know if you need help getting that set up.
Before we get to that though, I have to admit, I don't fully understand what this macro does. Even the doc example did not clarify that for me. 
Couple of thoughts:
- is `relative_datediff` the best name to signify what this does? Relative to what?
- you mentioned you're using this for lifetime analysis, what is the exact application there for this macro? (the `months_since_acquisition` example makes it seem like the benefit is this is 0-based, which doesn't seem the main benefit).
- could we add another example to the docs? 
- how does this work when `datepart` is less than a day, e.g. `hour`?
- how does this work when `datepart` is `week`, or worse, `isoweek`?

File: README.md [hunk 2]
```diff
@@ -766,3 +770,19 @@ or, optionally, you can override the default timezone:
...
 ```
+
+### [relative_datediff](macros/utils/relative_datediff.sql)(`date1, date2, datepart`)
+
+A date diff macro that can account for non calendar computed diffs between two date like fields, `date1` and `date2`, measured in `datepart`.
+
+When `day` is passed as the `datepart`, this reduces to the usual dbt utils datediff.
+
+For example,
+`{{dbt_date.relative_datediff("'2021-11-05'", "'2021-12-02'", 'month')}}` returns 0 while
```
Comment (clausherther) [inline README.md:781]: Couple of small nits (1):
`{{dbt_date.` should be
`{{ dbt_date.` (add space b/w `{{` and `dbt_date`). 
Same for next line.

File: README.md [hunk 3]
```diff
@@ -766,3 +770,19 @@ or, optionally, you can override the default timezone:
...
+`{{dbt_utils.datediff("'2021-11-05'", "'2021-12-02'", 'month')}}` returns 1.
+
+Ordering is as follows: if date2 is after date1, the returned datediff is positive.
+
+```sql
+    {{ relative_datediff('acquisition_date', 'order_date', 'month') }} as months_since_acquisition
```
Comment (clausherther) [inline README.md:787]: this should be
`{{ dbt_date.relative_datediff(`

File: macros/utils/relative_datediff.sql [hunk 4]
```diff
@@ -0,0 +1,13 @@
+{% macro relative_datediff(date1, date2, datepart) %}
+    {{
+        dbt_utils.datediff(
+            date1,
+            dbt_utils.dateadd(
+                'day',
+                '1 - ' ~ dbt_date.date_part('day',date1),
```
Comment (clausherther) [inline macros/utils/relative_datediff.sql:7]: Nit (2)
Missing space between `'day'` and `date1`, should be
`date_part('day', date1)`

**Comment by clausherther** [comment]
Hi @jpmmcneill just checking in if you had thoughts on my comments above?

**Comment by jpmmcneill** [comment]
Hey @clausherther - I haven't yet to be honest, but I'll get back to you soon!!!

Initial reaction - it could definitely be named better 😛

**Comment by jpmmcneill** [comment]
Hey @clausherther - I'm happy to make an integration test. I should be able to set it up myself, but if I run into difficulty I'll ask for help 😅

Answers:
> - is `relative_datediff` the best name to signify what this does? Relative to what?

It's basically relative to the original date. In some sense it could be called "calendar less". It's designet to have a date part agnostic method for modular arithmetic on the two dates. For example, if you're using years it would roughly work as: datediff (days)% 365. 

> - you mentioned you're using this for lifetime analysis, what is the exact application there for this macro? (the `months_since_acquisition` example makes it seem like the benefit is this is 0-based, which doesn't seem the main benefit).

I'd consider it quite useful for anything subscription based etc. It's basically just a tool that business analysts might use as a different lens on customers.. ie X calendar weeks old Vs X non calendar weeks old.

> - could we add another example to the docs? 

absolutely

> - how does this work when `datepart` is less than a day, e.g. `hour`?
> - how does this work when `datepart` is `week`, or worse, `isoweek`?

I don't think this macro makes sense with < 1 Day datepart. For week or isoweek, I'd expect roughly the same behaviour as there are 7 days per unit in both of those

**Comment by clausherther** [comment]
Hi @jpmmcneill! I'm going to close this one since it's a bit stale by now, but feel free to reopen if you want to push this one over the finish line. Thanks!

**Pull request closed without merging**